test/onnx/gen_onnx.py

```diff
@@ -2009,6 +2009,751 @@ def dynamicquantizelinear_2d_test():
+
+
+@onnx_test()
+def einsum_row_sum_test():
+    x = helper.make_tensor_value_info('x', TensorProto.FLOAT, [2, 3])
+    y = helper.make_tensor_value_info('y', TensorProto.FLOAT, [2])
+
+    node = onnx.helper.make_node('Einsum',
+                                 inputs=['x'],
+                                 outputs=['y'],
+                                 equation='ij->i')
+
+    return ([node], [x], [y])
+
+
+@onnx_test()
+def einsum_matrix_vector_multiplication_test():
+    x = helper.make_tensor_value_info('x', TensorProto.FLOAT, [2, 3])
+    v = helper.make_tensor_value_info('v', TensorProto.FLOAT, [1, 3])
+    y = helper.make_tensor_value_info('y', TensorProto.FLOAT, [2, 1])
+
+    node = onnx.helper.make_node('Einsum',
+                                 inputs=['x', 'v'],
+                                 outputs=['y'],
+                                 equation='ij,kj->ik')
+
+    return ([node], [x, v], [y])
+
+
+@onnx_test()
+def einsum_matrix_matrix_multiplication_test():
+    x1 = helper.make_tensor_value_info('x1', TensorProto.FLOAT, [2, 3])
+    x2 = helper.make_tensor_value_info('x2', TensorProto.FLOAT, [2, 3])
+    y = helper.make_tensor_value_info('y', TensorProto.FLOAT, [2, 2])
+
+    node = onnx.helper.make_node('Einsum',
+                                 inputs=['x1', 'x2'],
+                                 outputs=['y'],
+                                 equation='ij,kj->ik')
+
+    return ([node], [x1, x2], [y])
+
+
+@onnx_test()
+def einsum_vector_dot_product_test():
+    x1 = helper.make_tensor_value_info('x1', TensorProto.FLOAT, [3])
+    x2 = helper.make_tensor_value_info('x2', TensorProto.FLOAT, [3])
+    y = helper.make_tensor_value_info('y', TensorProto.FLOAT, [1])
+
+    node = onnx.helper.make_node('Einsum',
+                                 inputs=['x1', 'x2'],
+                                 outputs=['y'],
+                                 equation='i,i->')
+
+    return ([node], [x1, x2], [y])
+
+
+@onnx_test()
+def einsum_matrix_dot_product_test():
+    x1 = helper.make_tensor_value_info('x1', TensorProto.FLOAT, [2, 3])
+    x2 = helper.make_tensor_value_info('x2', TensorProto.FLOAT, [2, 3])
+    y = helper.make_tensor_value_info('y', TensorProto.FLOAT, [1])
+
+    node = onnx.helper.make_node('Einsum',
+                                 inputs=['x1', 'x2'],
+                                 outputs=['y'],
+                                 equation='ij,ij->')
+
+    return ([node], [x1, x2], [y])
+
+
+@onnx_test()
+def einsum_hadamard_product_test():
+    x1 = helper.make_tensor_value_info('x1', TensorProto.FLOAT, [2, 3])
+    x2 = helper.make_tensor_value_info('x2', TensorProto.FLOAT, [2, 3])
+    y = helper.make_tensor_value_info('y', TensorProto.FLOAT, [2, 3])
+
+    node = onnx.helper.make_node('Einsum',
+                                 inputs=['x1', 'x2'],
+                                 outputs=['y'],
+                                 equation='ij,ij->ij')
+
+    return ([node], [x1, x2], [y])
```

Comment:
How is this currently parsed? The simplest parsing would be an element-wise `mul`, but I don't think that's how it currently works. I'm guessing it's current using `dot` with something like: `C_ij = A_ij B_ij = A_ij1 B_j1i`?

Reply:
At the moment, we would do the following for inputs of shape `{2, 3}`:
```
@8 = ref::reshape[dims={6, -1, 1}](@6) -> float_type, {6, 1, 1}, {1, 1, 1}, target_id=0
@9 = ref::reshape[dims={6, -1, 1}](@7) -> float_type, {6, 1, 1}, {1, 1, 1}, target_id=0
@10 = ref::transpose[permutation={0, 2, 1}](@9) -> float_type, {6, 1, 1}, {1, 1, 1}, target_id=0
@11 = ref::dot(@8,@10) -> float_type, {6, 1, 1}, {1, 1, 1}, target_id=0
@12 = ref::reshape[dims={2, 3}](@11) -> float_type, {2, 3}, {3, 1}, target_id=0
```
So basically a batch matmul with 6 batches and 1x1 matrices.

Reply:
OK, that's fine. We should make an issue to track that this can be simplified by a later compiler pass to just a `mul` operator. If there are any other known simplifications that can be done at a later date we should also have those in an issue. This way we don't have to rediscover those simplifications later on.

Reply:
A more general approach I did here to convert skinny matrices to mul+reduce:

https://github.com/ROCm/AMDMIGraphX/compare/develop...skinny-dot-2-reduce

Although the dot -> mul directly should probably be done in simplify_algebra.